File: app/src/main/java/com/inguana/weatherapp/network/repositories/WeatherRepository.java
package com.inguana.weatherapp.network.repositories;

import androidx.lifecycle.MutableLiveData;

import com.inguana.weatherapp.model.Area;
import com.inguana.weatherapp.network.WeatherApiClient;

public class WeatherRepository {

    private static WeatherRepository instance;
    private static WeatherApiClient weatherApiClient;

    public static WeatherRepository getInstance() {
        if(instance == null) {
            instance = new WeatherRepository();
        }
        return instance;
    }

    private WeatherRepository() {
        weatherApiClient = WeatherApiClient.getInstance();
    }

    public void searchWeather(String searchQuery) {
        weatherApiClient.searchWeather(searchQuery);
    }

    public MutableLiveData<Area> getArea() {
        return weatherApiClient.getArea();
    }

    public MutableLiveData<String> getErrorMessage() {
        return weatherApiClient.getErrorMessage();
    }
}

File: app/src/main/java/com/inguana/weatherapp/network/repositories/WeatherRepository.kt
package com.inguana.weatherapp.network.repositories

import androidx.lifecycle.MutableLiveData
import com.inguana.weatherapp.model.Area
import com.inguana.weatherapp.network.WeatherApiClient

class WeatherRepository private constructor() {
    init {
        weatherApiClient = WeatherApiClient.instance
    }

    fun searchWeather(searchQuery: String?) {
        weatherApiClient!!.searchWeather(searchQuery)
    }

    val area: MutableLiveData<Area?>?
        get() = weatherApiClient?.area
    val errorMessage: MutableLiveData<String>?
        get() = weatherApiClient?.errorMessage

    companion object {
        var instance: WeatherRepository? = null
            get() {
                if (field == null) {
                    field = WeatherRepository()
                }
                return field
            }
            private set
        private var weatherApiClient: WeatherApiClient? = null
    }
}